language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -77,45 +68,116 @@ import           Data.HashMap.Monoidal
 import qualified Data.Map                                           as Map
 import           Data.Text.Prettyprint.Doc
 
+import           Data.Array
+
 {- Note [Scoping]
-The CEK machine does not rely on the global uniqueness condition, so the renamer pass is not a
+   The CEK machine does not rely on the global uniqueness condition, so the renamer pass is not a
 prerequisite. The CEK machine correctly handles name shadowing.
 -}
 
+type TermWithMem uni = WithMemory Term uni
+type TypeWithMem uni = Type TyName uni ExMemory
+type KindWithMem = Kind ExMemory
+
+-- 'Values' for the modified CEK machine.
+data CekVal uni =
+    -- TODO: we probably want to store a @Some (ValueOf uni)@ here, but then we have trouble in
+    -- 'readKnownCek'. I'll reconsider the way we deal with annotations once again.
+    VCon (TermWithMem uni)
+  | VTyAbs ExMemory TyName KindWithMem (TermWithMem uni) (CekValEnv uni)
+  | VLamAbs ExMemory Name (TypeWithMem uni) (TermWithMem uni) (CekValEnv uni)
+  | VIWrap ExMemory (TypeWithMem uni) (TypeWithMem uni) (CekVal uni)
+  | VBuiltin
+      ExMemory
+      (CekValEnv uni)   -- Initial environment, used for evaluating every argument
+      StagedBuiltinName
+      Int               -- Number of arguments to be provided (both types and terms)
+      [TypeWithMem uni] -- The types the builtin is to be instantiated at.
+                        -- We need these to construct a term if the machine is returning a stuck partial application.
+      [CekVal uni]      -- Arguments we've computed so far.
```

Comment:
I think if we did this it would make several things simpler: we could have a single function that takes a new argument, decides if it's got all of them, and tries to evaluate the builtin. Then we call that from `applyEvaluate` and `instantiateEvaluate`, which would handle some of my comments in the latter.

As a bonus, I think at that point we could probably inline both those functions.

Reply:
> Shouldn't this be one list of potentially-interleaved type and term arguments?

That's discussed in PR #2208.  Possibly, but that introduces some inefficiency in a critcal part of the evaluator, and (a) it's trapping something that the typechecker will object to anyway, (b) if we do hand something bad to the constant application machinery then we'll get an error.  Also, the idea of allowing higher-rank builtins has only really come to the fore in the last few days and it's still not clear if we want them.

Reply:
> (a) it's trapping something that the typechecker will object to anyway

Wait, why?

Reply:
> > (a) it's trapping something that the typechecker will object to anyway
> 
> Wait, why?

Isn't that right?  I think the problem is if we have a builtin `bn` of type  `int -> ∀t. (int -> int)` say then `[ { [(builtin b) (con 4) ] (con integer)} (con 5} ] is OK, but if we just count the type and term arguments in the machine and discard the types, we could equally well give it something with the arguments incorrectly interleaved like  `[ { (builtin b) (con integer) } (con 4)] (con 5} ] ` and it would still evaluate correctly.  If we run the typechecker it should surely accept the first but not the second.  We could replace the arity with a list of `termArg | typeArg` values and then the evaluator would reject it as well.  I'm coming round to the idea that we should do that, but I don't think it's essential if we believe that the code's been typechecked (and mybe even if it hasn't been).

Reply:
Ah, thank you for the clarification, I just completely misunderstood what you were saying. [I think](https://github.com/input-output-hk/plutus/pull/2208#discussion_r466683931) we should carefully handle the type/term distinction even where there are no types anymore.